operator_rolling_upgrades.md

```diff
@@ -0,0 +1,54 @@
+# Operator Rolling Upgrades
+Optionally allow a Infinispan CR cluster to be upgraded with zero downtime and no loss of data.
+
+## Spec
+Add a field to the CR spec that determines whether an upgrade should be `rolling` or `gracefulShutdown`. The latter is the current approach that does not allow data migration.
+
+```yaml
+apiVersion: infinispan.org/v1
+kind: Infinispan
+metadata:
+  name: example-infinispan
+spec:
+  upgrades: rolling
+```
+
+## Procedure
+When an upgrade is possible, the operator performs the following for each Infinispan CR:
+
+1. Create a new StatefulSet for the target cluster based upon the CR spec, wait until the cluster is formed
+
+2. Retrieve all cache configurations from the cluster in the original Statefulset
+
+3. Append a remote-store configuration to each of the ^ configurations and create cache target cluster:
```

Comment:
@ryanemerson if we need separate configurations for source and target clusters? If yes, the ConfigMap's should be separate also. Or we need to embed this config on the cluster startup.

Reply:
I think using the same config-map should be fine until an operator upgrade results in a new config being required for the server, in which case we would need to create a new ConfigMap.  So I think it makes sense to future proof the initial design and create a new ConfigMap to cover this possibility.

Reply:
DNS query for JGroups is part of the ConfigMap and we will need a different one for a new cluster

Reply:
Very good point :sweat_smile: